### ISSUE TITLE
Last change time does not have time zone

### ISSUE THREAD

**Issue by xmlf** [fisharebest/webtrees#2497]
I have set both the website preferences and my account time zone to UTC+8 (Asia/Shanghai), but the last change time shown in the individual facts and events is UTC time.
I think it should be more reasonable to display according to the time zone setting of the website preferences.

**Comment by ric2016** [comment]
The last change time is stored in the Gedcom data - without a timezone (Gedcom does not support timezones, see e.g. #1120). It therefore cannot be converted reliably to a specific timezone.

**Comment by vytux-com** [comment]
To be honest the fact that gedcom doesn't support timezones is not an issue. You should always store the time in UTC and then convert on display to whatever TZ you want. If needed in import of data we can ask what original TZ was in the file just in case it was non standard UTC time

**Comment by ric2016** [comment]
> You should always store the time in UTC

I don't think this is possible in general: Webtrees allows to store a time for birth and death events (I just realized that this isn't even in the Gedcom 5.5.1 spec, so it's anyway problematic). This time shouldn't be stored as UTC, because you'd have to change the respective date in some cases (when local time is on previous/next day compared to UTC), which is definitely not desirable.

So this could at most be done for change dates (and the transmission date in the Gedcom header), i.e. for the locations within the Gedcom where TIME is actually supported by the spec. Even then, the change dates in existing Gedcom files will often not be consistent wrt the implied time zones.

**Comment by fisharebest** [comment]
> So this could at most be done for change dates (and the transmission date in the Gedcom header)

Corrrect.

The `DATE` and `TIME` fields are processed independently.  The date processing includes converting from Gregorian to other calendars.  Calendar conversions that contain times are difficult.  (Arabic and Hebrew calendars start each new day at local sunset).

So, this means that

* we must have special handling for `CHAN` facts.
* we cannot convert the timestamp to the user's prefered calendar.

FYI, there is an interesting (and very good) document [GEDCOM 5.5.1 Annotated Edition](https://www.tamurajones.net/DownloadGEDCOM551AnnotatedEdition.xhtml).  When this document was being created, I contacted the author and suggested that `HEAD/DATE/TIME` and `CHAN/DATE/TIME` should always be assumed to be in UTC.  My suggestion was not included in the final version.

**Comment by GreggShimokura** [comment]
Hi,  2.x release is very nice.

Related to TZ, when I was trying to get my father to use Webtrees, he was bothered by the fact that events shown in the "On this Day" block were 'wrong' because it was one day ahead.  We are in UTC-8 (Vancouver, Canada).

To check does "On this Day" use the UTC time of events somehow?  It seems that for "On this Day" only the day values should be used.

Are we seeing the expected behaviour and/or is there a setting we can change to have the "On this Day" events show up on the 'proper' day?

Thank you,

Gregg Shimokura

**Comment by xmlf** [comment]
You can change to UTC-8 time zone in my pages----my account.

**Comment by GreggShimokura** [comment]
Hi - very fast response.
Unfortunately that change did not seem to have the desired effect.
Server Setting is UTC-5, User Account is UTC-8
Today is Jan 5, On this Day shows Events for Jan 6.

As it is Jan 6 where UTC-0 is now, it seems On This Day is using UTC-0?

Gregg

**Comment by xmlf** [comment]
Site preferences, also want to set.
http://xxx.com/admin/admin-site-preferences

**Comment by GreggShimokura** [comment]
Yes thanks
http://xxx.com/admin/admin-site-preferences
Server Setting is UTC-5
My pages/My account
User Account is UTC-8

Changing 
http://xxx.com/admin/admin-site-preferences
to  UTC-8
-> Did not change anything.

The physical server is located at UTC-5....

Gregg

**Comment by xmlf** [comment]
change the physical server to UTC-8

**Comment by GreggShimokura** [comment]
Hi I can't do that. 

The server is hosting other services.

Gregg

**Comment by xmlf** [comment]
https://github.com/fisharebest/webtrees/tree/master/modules_v4/example-server-configuration.disable
You can create a new module, change it here.
or use `.user.ini`

**Comment by GreggShimokura** [comment]
Hi Again.

To clarify, are you saying that I should make a new "On This Day" module?

Or a module to do something else like change the server time?

I am not sure I am equipped to do either...

Gregg

**Comment by xmlf** [comment]
It is very simple.
Rename this folder, remove disable,
And modify the corresponding parameters according to `ExampleServerConfigurationModule.php`.
https://github.com/fisharebest/webtrees/issues/2308

If. user.ini can be used, it is recommended to use. user.ini.
https://www.gearhost.com/documentation/how-to-configure-user-ini
https://www.pair.com/support/kb/php-ini-vs-user-ini/

**Comment by fisharebest** [comment]
@GreggShimokura - this issue is about the GEDCOM `CHAN` records, and how they do not support time zones.

You are talking about something totally different.  Internally, webtrees uses UTC, and is converted for each user based on their time zone.  Server settings should have nothing to do with it.  If your uncle has the right time zone set in his account settings, then can you create a new/separate issue for this?

Discussing two different things together just causes confusion.  For example, issues can only have one status (open/closed), the issue number is used to track code changes, issues can be assigned to individuals, etc.

**Comment by ric2016** [comment]
@xmlf why did you close this now - the original issue has not been resolved!

**Comment by GreggShimokura** [comment]
Hi.. I can see how changing the TZ to UTC-8 for the webtrees directory might work for people in that TZ but how would it work for others say closer to UTC-0?

Gregg

**Comment by GreggShimokura** [comment]
I did try putting a .user.ini in the ../webtrees directory with one statement setting the TZ to UTC-8.
`date.timezone = "America/Vancouver" `

It did not have the desired effect..

Maybe we need a new issue related specifically to "On this Day"?

**Comment by vytux-com** [comment]
What timezone is configured for that user? In the user administration screen

**Comment by GreggShimokura** [comment]
Hi
User Administration: UTC-8 (America/Vancouver)

PHP.ini: UTC-5 (America/Toronto)
(cannot change)

webtrees/.user.ini: UTC-8 (America/Vancouver)

Gregg

**Comment by fisharebest** [comment]
> I did try putting a .user.ini in the ../webtrees directory with one statement setting the TZ to UTC-8. `date.timezone = "America/Vancouver"`

This will be ignored.  webtrees works internally in UTC.  
The first thing it does is set the PHP and MySQL timezones to UTC.

When we **display** datetimes, we convert to the timezone of the user (when logged in) or the timezone of the size (for visitors).

> Maybe we need a new issue related specifically to "On this Day"?

Yes.

**Comment by GreggShimokura** [comment]
"On this Day" issue created: [#2961](https://github.com/fisharebest/webtrees/issues/2961)

**Comment by TheDutchJewel** [comment]
> When we **display** datetimes, we convert to the timezone of the user (when logged in) or the timezone of the size (for visitors).

Strange. I am logged in to my localhost and my time zone (Amsterdam) is currently (daylight saving time) differing 2 hours from UTC. This is well displayed on my pc but not in webtrees: the last change shows the time in UTC.

Does this correspond to the quoted statement?

**Comment by fisharebest** [comment]
> Does this correspond to the quoted statement?

Yes.  The `CHAN:DATE` and `CHAN:DATE:TIME` fields do not have a time-zone.

**Comment by fisharebest** [comment]
GEDCOM allows us to store dates and times.

```
1 BIRT
2 DATE 20 JAN 1999
3 TIME 15:23
1 CHAN
2 DATE 20 JUL 2026
3 TIME 11:03:32
```

GEDCOM does not allow us to store time-zones.

Date/time for genealogy events (e.g. `BIRT`) is assumed to be local time.  We don't know the location.  It is just displayed as descriptive text.

Date/time for metadata (e.g. `CHAN`) is created as UTC.  Imported metadata has an unknown timezone.

For CHAN timestamps, we have two choices.

1) ignore timezone, because we don't know the timestamp of imported data
2) treat as UTC, because all data created by us is known to be UTC.

I will implement (2)
